pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+package model
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/segmentio/ksuid"
+)
+
+// ConflictMode indicates the conflict detection mode defined for a diamond
+type ConflictMode string
+
+const (
+	// IgnoreConflicts is the diamond mode with which conflicts are not handled (latest win, no track kept of clobbered files)
+	IgnoreConflicts ConflictMode = "ignored"
+
+	// EnableCheckpoints is the diamond mode with which conflicts are explicitly handled and saved as "checkpoints" (incremental upload)
+	EnableCheckpoints ConflictMode = "enable-checkpoints"
+
+	// EnableConflicts is the diamond mode with which conflicts are detected and saved as "conflicts"
+	EnableConflicts ConflictMode = "enable-conflicts"
+
+	// ForbidConflicts is the diamond mode with which conflicts result in failure to commit
+	ForbidConflicts ConflictMode = "forbids-conflicts"
+)
+
+// DiamondState models the running status of an ungoing diamond workflow
+type DiamondState string
+
+const (
+	// DiamondInitialized is the state of an initialized diamond
+	DiamondInitialized DiamondState = "initialized"
```

Comment:
more on expressing language constructs in golang, `enum` edition:

the only thought i've put into `enum` types so far is that we use `iota` to make `enum`s in golang a lot more like `enum`s in C wherein we're declaring integer constants rather than `enum`s in Java where we have something that's a lot more like an algebraic data type except without parameters on any of the unioned record types.

**succinctly**
since the preceding paragraph is probably meaningless, what i'm saying specifically is to `DiamondInitialized DiamondState = itoa` here, then `switch` from C-style enum (just an `int`!) to string value for serialization.

this is not based on anything other than preference and may be safely disregarded if it matters too much one way or another, possibly from a perspective i'm not considering.

Reply:
iota is just a syntactic sugar to avoid repeating numerical indices.
Both idioms exist in goland and this doesn't change the way switch statements look like.